src/clj/puppetlabs/services/jruby/jruby_metrics_core.clj

```diff
@@ -298,9 +379,24 @@
    :lock-requests (atom {})
    :lock-status (atom {:current-state jruby-pool-lock-not-in-use
                        :last-change-time (timestamp)})
+   :metric-registry registry
+   :hostname hostname
```

Comment:
We could move all of these timers into their own registry, but there are a couple of complicating factors:

  - We'd have to also put the existing `borrow-timer` into the new registry in order to keep the full set together.

  - We'd have to copy in logic for wiring the registry into JMX and the Graphite reporter (if configured).

By using the existing registry we get the second point for free from the TK init routines, which saves a lot of work.

Reply:
Yeah, that's a good point about not creating an additional registry that folks would have to configure.

I think that the surrounding code assumes objects that can be initialized and then acted upon, queried, for better or worse, though.

What if, instead of saving the hostname and registry into the `metrics` map, we created an atom that contained them and the existing `borrow-timer` metric. Then we can create an "update-borrow-timers" kinda function that takes the `borrow-timers` atom and the `TimestampedReason` and updates the total and per-reason timers (like lines 274-277 in `track-return-instance!`). We can then also just pass that atom into `summarize-borrow-timers` instead of the whole metrics map. And per https://github.com/puppetlabs/puppetserver/pull/1689/files#r189341818 basically get rid of `top-borrow-summaries`.

I don't know if that'd be worth the effort, but I'm not a huge fan of passing all of metric through `top-borrow-summaries` into `summarize-borrow-metrics` or adding `hostname` and `registry` to the metrics map.

Reply:
After hemming and hawing about this quite a bit, I don't think it's worth the extra effort. `hostname` and `registry` have to be in the metrics map in order to dynamically generate new timers. Adding an atom could remove the need for the `borrow-timers` method and memoize the overhead of filtering the metrics list. But, it would push complexity into other functions and we already do several similar filter operations per-status call over in the `puppet-profiler` service.

This definitely could be done better, but I think:

  - It's good enough at the moment.
  - The best approach to eliminate state like `hostname` may be to improve the encapsulation that the `trapperkeeper-metrics` service offers.